### DISCUSSION TITLE
[Feature] Enhancing face id Correlation for Video

### DISCUSSION THREAD

**Discussion by zacharyvmm** [immich-app/immich#5936] [Feature Request]
### The feature

In my observations, it appears that the correlation with face id is established based on the first frame of the video. While this is a straightforward and easily implementable solution, there are some potential issues.

Notably, if the video doesn't commence with a person, there could be various reasons for this, including the camera being out of focus or the video starting by showcasing the floor (as is common in many family videos).

To address this, two potential solutions come to mind:
1. Generate [multiple thumbnails](https://superuser.com/questions/538112/meaningful-thumbnails-for-a-video-using-ffmpeg) for each video.
2. Allow users to manually associate videos and images in cases where the AI doesn't automatically pick up on the relevant content.

Ideally, the implementation would incorporate both solutions.

### Platform

- [X] Server
- [X] Web
- [ ] Mobile

**Comment by mertalev** [comment]
We're considering (1) for both smart search and facial recognition. The advantages you've raised apply to both. Additionally, for smart search, it would mean you could search for a specific scene in a video.

Another possible feature could be to do machine learning on-the-fly when pausing a video, showing the detected faces for that frame. This could also be used to add OCR for a frame once that machine learning task has been added.

**Reply by adasium** [reply]
I think (2) is equally as important, if not more because:
- AI might not detect faces
    - I assume this can happen no matter the frame so I'd be left with missing faces and I don't want to meddle with face detection settings or searching for the split frame the person was visible or rerun jobs just to tag one photo when manual fix would be faster
- I'd like to tag people that are not in the photo (for example the cameraman or people I was with when the video was taken but are not included in the frame)
    - this would make searching for specific photos easier

**Comment by DaanHend** [comment]
Would be a nice additition to have facial recognition work on videos 💯 

#7953

**Comment by aviv926** [comment]
See also corresponding thread on Discord:
https://discord.com/channels/979116623879368755/1222898437490479104

**Comment by Azuraell** [comment]
This feature would be amazing on Immich. I am currently searching for pictures of a deceased cherished relative, and it is so much work. Also I know I am missing a lot because I can't watch every single video.

**Comment by tit1** [comment]
One approach to doing this and IMO may be fairly simple is to use ffmpeg to extract screen shots from the video, and run those through your same systems you have in place for face recognition and smart search. The images could be removed after or saved as a stack of reference images for the video.  The tricky part is determining the time interval for grabbing images. You could have this as as system setting changed by the user or a fixed rate that could be changed in the UI video by video.

**Reply by mrAndream** [reply]
**Add face detection for entire videos:**

- Extract frames every X seconds (configurable).
- Detect faces using existing models.
- Cluster same faces into one person.
- Cluster same faces timing into one time slots.
- Save timestamps and coordinates.
- Delete extracted frames after processing.

**Notes**
Use FFmpeg for frame extraction.
Extend existing face recognition system.

**Comment by austinbaccus** [comment]
Any updates on this?

**Reply by marvil6** [reply]
I would love this feature as well.

**Comment by TKaluza** [comment]
Maybe this is interesting:
https://arxiv.org/pdf/2502.03183

**Reply by bo0tzz** [reply]
That's a very neat approach! Do I understand correctly that it still samples frames by count or keyframe at the start, and so avoids generating an unused embedding for all/many frames?

**Reply by DaanHend** [reply]
Interesting!

It seems like it creates vectors for each frame, then picks the ones that best represent the whole 'vector timeline.' Those key frames could be really useful for ML in Immich. Seems more efficient than processing every single frame with ML. If that vector conversion would be simple then...

**Reply by bo0tzz** [reply]
> It seems like it creates vectors for each frame

That's not my impression of the paper. Doing that would be very slow because generating the vector embedding is expensive.

**Reply by mertalev** [reply]
It does generate embeddings with CLIP first. I think the point of that paper is to optimize for fewer frames passed to the VLLM, which is much slower than CLIP. This apparently increases the VLLM result quality as well. For our use, if you sample one frame every second and generate an embedding for it, then you're already done without needing anything in this paper.

**Comment by ireun** [comment]
I think what is needed for Immich has already been implemented for Jellyfin - it is called Trickplay, and it is used to help with seeking timeline. It creates a file with frames every x seconds, I think such image could be passed into the face detection/recognition.

Here is how it looks:
![0](https://github.com/user-attachments/assets/8f2b39cb-a1f4-49ed-b2a7-26e57833889f)


You may want to look at this (and related) commit https://github.com/jellyfin/jellyfin/commit/ca7d1a13000ad948eebbfdeb40542312f3e37d3e

https://deepwiki.com/search/explain-trickplay_863319de-15e3-4cc6-83d4-048c6773bff2

**Reply by tylers-username** [reply]
There's something more to consider—voice detection. 

Often it's, in this example, Mom is or Dad holding the phone talking to the subject of the video but you never see them. The sentimental value and job to be done here, is that let's say Mom or Dad passes, you'd be able to find videos of them that would have otherwise remained buried.

**Reply by ireun** [reply]
I've looked at jellyfin implementation and it seems to be pretty simple all through ffmpeg:

1. have a frame every N-seconds (https://ffmpeg.org/ffmpeg-filters.html#fps-1) simply `fps=1/N`
1.1. Or maybe have N frames for a video? (0:00:00, the end, and N-2 in between) though I think ^^^^ would be better
3. generate JPEGs using image2 encoder (https://ffmpeg.org/ffmpeg-formats.html#image2_002c-image2pipe)
4. Feed JPEGs to ML buffalo or what not

Example: `ffmpeg.exe -i .\35f4037bcf569a1c63d7aeed2faf744c.mp4 -vf fps=1/1 -f image2 "out%08d.jpg"`

More work would be needed probably to prepare api/model/ui to make it work nicely (show only one face for a person in UI? maybe lower max-distance for matching faces coming from one video?)

**Reply by DankMemeGuy** [reply]
> There's something more to consider—voice detection.
> 
> Often it's, in this example, Mom is or Dad holding the phone talking to the subject of the video but you never see them. The sentimental value and job to be done here, is that let's say Mom or Dad passes, you'd be able to find videos of them that would have otherwise remained buried.

I think this is unrelated to what this discussion is about. But using Whisper to transcribe audio so that Smart Search can allow you to search words would help, and I think you might be able to identify voices using Diarization: https://www.youtube.com/watch?v=aC4LYberXys


That all being said, the issues are this:

1. Face recognition only occurs for the first frame -> problem that requires every frame (or every nth) to be scanned
2. Videos don't have much scanning for anything deep -> problem that requires scanning so you can search the transcript, scene, speaker

The two aren't distinctly related.

**Comment by unpleased** [comment]
Any update on this? It's been years

**Reply by SansGuidon** [reply]
@unpleased there is a discount on pull requests currently on GitHub. You can submit your solution to this years long problem, for free, it will barely cost you anything especially if using LLM.

**Reply by bo0tzz** [reply]
(please don't open a PR generated entirely with an LLM >.<)

**Reply by devasheeshG** [reply]
Hey, if you’d like, I can take a look at this issue. Can you confirm that no one else is working on it right now?

**Reply by devasheeshG** [reply]
> @unpleased there is a discount on pull requests currently on GitHub. You can submit your solution to this years long problem, for free, it will barely cost you anything especially if using LLM.

btw he was just asking if there has been any update or not on this feature

**Reply by bo0tzz** [reply]
@devasheeshG as long as you're the same person that was asking about this feature on Discord today then you're the only one, go for it!

> btw he was just asking if there has been any update or not on this feature

If there are any updates it'll be plenty clear from discussion on this thread; if there are none, asking for "any updates" unnecessarily notifies everyone subscribed to this thread.

**Reply by devasheeshG** [reply]
> @devasheeshG as long as you're the same person that was asking about this feature on Discord today then you're the only one, go for it!

no, i was not asking for any updates on discord.

ok then, i'll look into it, draft a solution, and share a summary here, then you can confirm if that's good, then i'll just go ahead and implement that

> If there are any updates it'll be plenty clear from discussion on this thread; if there are none, asking for "any updates" unnecessarily notifies everyone subscribed to this thread.

ig he was thinking maybe this is already implemented bts; i also thought so initially lmao

**Reply by eWOOD29** [reply]
> > @devasheeshG as long as you're the same person that was asking about this feature on Discord today then you're the only one, go for it!
> 
> 
> 
> no, i was not asking for any updates on discord.
> 
> 
> 
> ok then, i'll look into it, draft a solution, and share a summary here, then you can confirm if that's good, then i'll just go ahead and implement that
> 
> 
> 
> > If there are any updates it'll be plenty clear from discussion on this thread; if there are none, asking for "any updates" unnecessarily notifies everyone subscribed to this thread.
> 
> 
> 
> ig he was thinking maybe this is already implemented bts; i also thought so initially lmao

Any progress on implementing this? I would love this feature.

**Reply by bo0tzz** [reply]
@eWOOD29 that question was already asked and answered above. https://justinmayer.com/posts/any-updates/

**Reply by ireun** [reply]
@bo0tzz maybe it's time to start marking some comments here as 'off-topic'

**Comment by 0thomasholland** [comment]
Hey, I've picked this up and have started working on it, hope that's okay as there hasn't seemed to be much movement on this for a little while and I would quite like to use this feature!

The approach broadly follows what @ireun outlined, passing each frame (wrt. grid system I am unsure how it would affect the model certainty outputs and doing framewise means I don't have to worry about frame placement within merged tile photo for timestamp extraction and/or placement on the video during paused frames etc.; downside is more compute). 

Doing this using ffmpeg's `fps=1/N` filter to sample frames at a configurable interval, then feeding each frame through the existing face detection pipeline. I've got the frame extraction utility working and am planning to break the rest up roughly like this:

1. (done) Frame extraction utility and two new config options for frame interval and max frames per video.
2. New jobs to detect faces across the sampled frames and store results with a timestamp so we know where in the video each face appeared.
3. A deduplication step to cluster faces from the same video before they reach the recognition pipeline to avoid flooding it with near-identical frames of the same person. Thinking greedy cosine clustering using the same max distance threshold already used for recognition, to avoid introducing another tuneable? Happy to hear if there's a preferred approach here? (This helps negate the need for grid tiles system.)
4. Include the timestamp in the face API response and add a new endpoint returning the first timestamp a person appears in each video as could be useful/used for seeking directly to them in a long recording.
5. Manual job trigger and the new frame-sampling settings wired into the admin UI.
6. Look at how to surface this to the user. Again, please say if there's a preferred approach here?

Something that could be looked into later on is a branching search for the first frame appearance (if first apears at `n`, and previous frame was `m` then look at midpoint frame `[n+m]/2` and if the face is present in midpoint, then do the same midpoint between that frame and `m` and if not then midpoint frame and `n` then repeat until frame is found).

I don't have the hardware (I'm a student who has just enough hardware to run my live immich server) to properly test this against large libraries or long videos so I would really appreciate any help testing once things are further along. Also, I would want to keep this manageble for now, so wouldn't be looking at the voice nor off-camera people tagging if that was okay?

Thanks! Any feedback or suggestions welcome.

**Reply by bo0tzz** [reply]
cc @mertalev

**Reply by marvil6** [reply]
Sounds great! I'd love to be a tester if you need it.

**Reply by 0thomasholland** [reply]
Small update - have done the first two items of the list

Also made a python script to make sure the image processing logic and workflow through the ML worked (I'm more used to python so was quicker for me to quickly script it up) ... prelim results look okay:

https://gist.github.com/0thomasholland/b73211bd63fdf668b8000beb4fa59b02 (scroll to bottom for output)

**Reply by mertalev** [reply]
The discussion around ML processing in videos has revolved around a related subject: trick play generation. I think it'd be good to align these features so they can build on each other. It's also worth noting that #28023 adds keyframe metadata to the DB that can be useful here (for example, using `-skip_frame nokey` is much more efficient than `fps`, but you need to know whether the video's keyframe interval is short enough for this to work well).

**Reply by JMSwag** [reply]
First let me say thanks for taking this up. A possible optimization for the key frame/N frames step is to use scene detection to segment the video based on scene changes, then grab a thumbnail from each segment to run through CLIP.

**Reply by ireun** [reply]
umm.. scene changes? I believe such feature is mostly intended in personal videos - shot using single camera. There are usually not much of a distinction in terms of scene.

**Reply by JMSwag** [reply]
@ireun Maybe not in the literal movie/tv definition (shot change) but a significant change across time. For example, a person/landmark entered or left the scene.

**Reply by 0thomasholland** [reply]
@mertalev - thanks for flagging #28023. I've implemented two versions my initial version which uses 1/N with sanity checks (length of video namely for max frames to process) and another based off this. For the metadata I've done it so frame extraction checks avg keyframe interval (`totalDuration / keyframeCount`), uses `-skip_frame nokey` when interval is short enough, falls back to `fps=1/N` otherwise. Also kept the extraction utility generic so trickplay can build on the same code. Was there anything else you would want to look to include from this? 

I've kept the branch with the metadata integration separate for now so that it can be merged independently of the face recognition work, but I can merge them if you think it would be better to do it all at once (I have each stage seperate so that each PR could be reviewed stage-wise for similar logic).

@JMSwag / @ireun : the within-video deduplication step addresses a similar goal: face embeddings from sampled frames are cosine-clustered before hitting the recognition pipeline, so near-duplicate frames of the same person are collapsed regardless of whether there's a hard scene cut. Avoids the external scene-detection dependency and (as @ireun notes) personal video rarely has sharp scene boundaries anyway.

It's in a stage where having others looking at it would be helpful, I can open PRs from my fork but also always happy discuss further before proceeding if you'd prefer? 

standalone version: is up to https://github.com/0thomasholland/immich/tree/feature/video-face-phase-5
from @mertalev / #28023 PR: https://github.com/0thomasholland/immich/tree/feature/video-metadata-integration

thanks all!

**Reply by mertalev** [reply]
Thanks for working on this! My thought is to extract keyframes with `-skip_frame nokey` in one process for trick play (which will be quite relevant soon since there's active work on HLS), then send those trick play images to ML for processing. This makes the features much more cohesive, avoids redundant extraction work and allows FFmpeg to be called once per video rather than per-frame.

There are some design decisions here for how trick play should work (Jellyfin uses [EXT-X-IMAGE-STREAM-INF](https://developer.roku.com/docs/developer-program/media-playback/trick-mode/hls-and-dash.md#hls-considerations), while the more HLS-native way is an [I-frame playlist](https://developer.apple.com/documentation/http-live-streaming/hls-authoring-specification-for-apple-devices#Trick-Play)) as well as data modeling. I also think this should ideally extend to other model tasks rather than coupling to face detection. CLIP in particular would be beneficial to run on frames, and an eventual IQA model should likewise run on them as well.

I think the best option is to open a focus topic on Discord so we can align on the direction for this feature and shape it into a design document.

**Reply by ireun** [reply]
I've looked at the commits in your branch @0thomasholland.
I still think there should be option that in a scenario: 
* cap is set to 50 frames,
* interval is set to 1 frame / 2 seconds,
* video length longer than `cap*interval` so let's say it is 10 minutes (600s) 

That it would. rather than, sampling first 100s (50 frames * 2s) and then ignoring the rest of video (500s), but to uniformly sample the video (1st frame at 0s, 50th frame at 600s and 48 frames in between (each 600/49 seconds from each other).

This would not be needed though if keyframes were used for this feature ( though as @mertalev pointed "but you need to know whether the video's keyframe interval is short enough for this to work well")

Also I wonder if ignoring the first/last few seconds of video would not give better results.

**Reply by 0thomasholland** [reply]
@mertalev thanks - have opened a topic on discord, good idea https://discord.com/channels/979116623879368755/1498660409924911206 . The metadata-integration branch already has most of the pieces for the unified extraction approach but just needs re-wiring once we've worked out what you want the shape of the consumer interface to look like. 

@ireun - yes I agree completely; the frame extraction implemented does this: https://github.com/0thomasholland/immich/blob/6283f6cfa1a5c701789298bc95d327a55dad1ca4/server/src/repositories/media.repository.ts#L310-L330 
I also have put in a check of if avg keyframe interval > 2× the desired frame interval (e.g. desired=2s -> threshold=4s), keyframes are too sparse so will fall back to -vf fps=1/N

**Reply by ireun** [reply]
Oh I had to miss that, thanks. Though there is no mention of this in the docs I believe?

**Reply by 0thomasholland** [reply]
Yes - I haven't done that bit yet actually, good spot thank you.  Have just updated the branch.

**Reply by IAfanasov** [reply]
@0thomasholland , I believe in u! U can do it!

**Reply by 0thomasholland** [reply]
sorry! in the middle of my master degree exams at the moment so a little slow with this! all the branches are here but (a) need testing and (b) need refining with others, esp wrt the shape of the consumer interface

https://github.com/0thomasholland/immich 

thanks for your patience!

**Reply by IAfanasov** [reply]
Hi @0thomasholland — I wanted this feature badly enough that I had an AI coding agent rebase your `feature/video-face-phase-5` onto current main and stress-test it. I only read immich's contributing guidelines afterwards and realized LLM-generated PRs aren't welcome — my mistake — so I won't be opening one. Sharing the results here instead, in case any of it saves you time. Zero obligation.

**Bugs it found in phase-5** (each reproduced/verified with tests, worth checking regardless of what you do with the rest — links pinned to `9aa5d9d`):

1. **Timestamps drift on capped videos** — `extractVideoFrames` widens the interval when `maxFrames` is hit ([media.repository.ts#L297](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/repositories/media.repository.ts#L297)), but `handleVideoDetectFaces` stamps `timestampMs` using the configured interval ([person.service.ts#L460](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/services/person.service.ts#L460)), so stored timestamps are wrong for any video long enough to hit the cap.
2. **No dedup against the preview face** — `getVideoFacesWithEmbeddings` filters `timestampMs IS NOT NULL` ([person.repository.ts#L133](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/repositories/person.repository.ts#L133)), so the preview-derived face never joins clustering and each person keeps ≥2 near-duplicate face rows.
3. **Re-runs append duplicates** — the manual job queues `data: {}` ([job.service.ts#L38](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/services/job.service.ts#L38)), so `force` is `undefined` and the `$if(force === false, ...)` freshness filter never applies ([asset-job.repository.ts#L251](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/repositories/asset-job.repository.ts#L251)); the previous run's faces aren't cleared before insert either ([person.service.ts#L485](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/services/person.service.ts#L485)).
4. **`videoFacesRecognizedAt` is never persisted** — the `ON CONFLICT` update map in `upsertJobStatus` doesn't include the new column ([asset.repository.ts#L284-L291](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/repositories/asset.repository.ts#L284-L291)), and the row always exists by then (the metadata job creates it first), so it stays `NULL` and "missing only" can't work.
5. **Preview re-detection deletes video faces** — `handleDetectFaces`' stale-face sweep collects every ML face ([person.service.ts#L339](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/services/person.service.ts#L339)) and removes the unmatched ones ([#L371](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/services/person.service.ts#L371)) — video faces never IOU-match preview detections, so any re-detection wipes them.
6. **Wrong person thumbnails** — when a video face becomes `person.faceAssetId`, `handleGeneratePersonThumbnail` crops the *preview image* with a bbox from a different frame ([media.service.ts#L423-L428](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/services/media.service.ts#L423-L428)); clustering deletes can also null `faceAssetId` via the FK's `SET NULL` ([person.table.ts#L58-L59](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/schema/tables/person.table.ts#L58-L59), [person.service.ts#L560](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/services/person.service.ts#L560)) with no repair path.
7. **Perf** — `fps=1/N` decodes the whole video to keep ≤50 frames ([media.repository.ts#L301](https://github.com/0thomasholland/immich/blob/9aa5d9ddf1217ccf2fcae9cad3e1924ea15b8b01/server/src/repositories/media.repository.ts#L301)); per-frame input seeks (`-ss`) make extraction cost independent of video length.

**The branch**: https://github.com/IAfanasov/immich/tree/feat/video-face-recognition — your 7 commits untouched at the base, then 4 commits fixing the above plus productionizing: opt-in config (default off, so upgrades don't reprocess existing libraries), a dedicated `videoFaceDetection` queue, a nightly sweep for incomplete videos, frame downscaling to preview size, person thumbnails re-extracted at the face's timestamp, web overlay gating, regenerated OpenAPI/TS/Dart artifacts, and unit + Postgres-backed tests. AI-written, but it passes the full test suites, the codegen drift gates, and a live ffmpeg+ML end-to-end run.

Cherry-pick, reimplement, or ignore entirely — all fine. Happy to answer questions, or delete the fork if you'd rather it not float around.

**Reply by ssjrocks** [reply]
EDIT**** this is also ai code, haha we both started working on this on the same day,i also worked on this feature using the work @0thomasholland  did as a base, https://github.com/ssjrocks/immich if anyone is interested, im doing alot of work on supporting ui features also, like tools for grouping and seperating detected faces and an "appearances" feature to view all a persons appearances in different videos, as well as better control over video sampling rates to allow for power users to literally scan multiple frames per second, with generous max frames settings, for example i have a family video that runs for 1.5 hours and i can scan the entire video at a scan rate of 2 frames per second, the result? i can see every occurance of any person in the video

**Comment by CarlitoGrey** [comment]
Not wanting to run a non-prod version of immich I decided to roll (with claude) a workaround which operates alongside immich taking advantage of the API. 

ImmichVideoFace 'ivf' scans the full video using keyframes plus at least a set number of frames per second (8 by default). It reuses embeddings from immich and runs face detection using the same model that generated the embeddings. 

Scanning and writing changes are separate, after each scan (can be batched) a HTML report is generated of the video and faces detected. Writes to immich have to be manually approved, at which point the person is added against the relevant video. Unknown faces are tracked over videos and can me promoted to a real person in immich. 

Note this has only been tested on my machine on one version of immich. Backup first. 

https://github.com/CarlitoGrey/ImmichVideoFace

**Reply by weslyth** [reply]
I'm ecstatic to see face detection in multiple video keyframes (real video face detection)! This could be a great revolutionary/milestone feature! Thank you for working on this much needed/awaited feature.
